text/metrics/0113-exemplars.md

```diff
@@ -0,0 +1,94 @@
+
+### Exemplar behaviour for standard aggregations
+
+#### HistogramAggregator
+
+Every bucket in the HistogramAggregator MUST (when enabled) maintain a list of exemplars whose values are within the boundaries of the bucket. Implementations should attempt to retain at least one exemplar per bucket, with a preference for exemplars with a sampled trace context and exemplars that were recorded later in the time period. They should also not retain an unbounded number of exemplars.
+
+#### Sketch
+
+A Sketch aggregation should maintain a list of exemplars whose values are spaced out across the distribution. There is no specific number of exemplars that should be retained (although the amount should not be unbounded), but the implementation should pick exemplars that represent as much of the distribution as possible. Preference should be given to exemplars with a sampled trace context. (Specific details not defined, see open questions.)
+
+#### Gauge
+
+Most (if not all) Gauges operate asynchronously and do not ever interact with traces. Since the value of a Gauge is the last measurement (essentially the other parts of an exemplar), exemplars are not worth implementing for Gauge.
+
+#### Exact
+
+The Exact aggregator does not aggregate measurements. If exemplars are enabled, implementations may attach a separate exemplar to each measurement in an exact aggregation including the trace context and full set of labels.
+
+Exemplars will always be retrieved from aggregations (by the exporter) as a list of Exemplar objects.
+
+## Trade-offs and mitigations
+
+Performance (in terms of memory usage and to some extent time complexity) is the main concern of implementing exemplars. However, by making recording exemplars optional, there should be minimal overhead when exemplars are not enabled.
+
+## Prior art and alternatives
+
+Exemplars are implemented in [OpenCensus](https://github.com/census-instrumentation/opencensus-specs/blob/master/stats/Exemplars.md#exemplars), but only for HistogramAggregator. This OTEP is largely a port from the OpenCensus definition of exemplars, but it also adds exemplar support to other aggregators.
+
+[Cloud monitoring API doc for exemplars](https://cloud.google.com/monitoring/api/ref_v3/rpc/google.api#google.api.Distribution.Exemplar)
+
+## Open questions
+
+- Exemplars usually refer to a span in a sampled trace. While using the collector to perform tail-sampling, the sampling decision may be deferred until after the metric would be exported. How do we create exemplars in this case?
+
+- We don’t have a strong grasp on how the sketch aggregator works in terms of implementation - so we don’t have enough information to design how exemplars should work properly.
+
+- The spec doesn't yet define a standard set of aggregations, just default aggregations for standard metric instruments. Since exemplars are always attached to particular aggregations, it's impossible to fully specify the behavior of exemplars.
```

Comment:
There is the added complication of the current push to have the default SDK support custom aggregations as well.

It might be useful to add a section above on how custom aggregations are left to their own to produce useful exemplars. And, possibly, how that would look.

Reply:
Agreed, exemplar support in custom aggregations would have to be rewritten for each aggregation if exemplars stay coupled to aggregations. We could pull exemplars out a level to function seperately from aggregations, letting the user plug in exemplar systems to custom aggregations (as well as define their own), I wrote a bit about it in a comment on the doc.

Reply:
> brings the scope of Exemplars outside of just the SDK into the API, which we might not want

Agreed, I don't think we want that.

Maybe it just goes without saying, that if you are writing a custom aggregator it's up to you to annotate appropriately with exemplars. Likely, it will only be the vendor or user who wrote the custom aggregation that will use it, so this is not an OTel issue.

Reply:
Using the proposed "sample_count" field described in a comment above, systems will be able to make meaningful interpretation of the exemplars (regardless of how they were selected). At the protocol level, I do not think we need to specify more than that.

This is a question about what the default SDK should be specified to do. There I think we can find some agreement:

- exemplars are disabled by default
- exemplars are enabled by indicating how many exemplars to collect per instrument
- turning on exemplars means asking for arbitrary selection (i.e., not using unbiased sampling)
- the histogram aggregator is expected to produce per-bucket exemplars when enabled, so you should configure exemplars in multiples of the number of buckets
- the protocol may support optional statistical count, but no default behavior takes advantage of it.

This sounds problematic. One of the problems is that it's not widely known how we can compute statistical samples. 

In https://github.com/open-telemetry/opentelemetry-specification/issues/608 there is a long discussion about creating a new instrument to capture a sum of individual measurements and a distribution of those measurements. Hypothetical name: CountRecorder. With statistically-sampled exemplars, we can get approximately the same output from a simple Counter. Let me explain how.

Let's say I use the Priority Sampling algorithm, because it's very simple to implement: https://dl.acm.org/doi/10.1145/1314690.1314696. 

The Counter instrument will be configured to collect 1000 exemplars using weighted sampling, with no more than 10 exemplars per distinct label set. Today's SDK has two distinct phases: in the Accumulator the aggregator combines events for a single label set, then the Integrator combines those results.

In the Accumulator (aggregating per label set), the Sum aggregator is used with an additional size=10 sampler to select exemplars using weighted sampling. Each `Add(N)` event on the aggregator will call the sampler with Weight=N. After the interval is complete, each Aggregator will contain <=10 exemplars. The sampler gives me an _output_ weight for each individual, and the computed `sample_count` described above is equal to `output_weight / N`, i.e., the ratio of output weight to input weight.  If there are fewer than 100 distinct label sets, we can simply use the exemplars at this stage because there are fewer than 1000 in total. If there are more than 100 distinct label sets, the exemplars must be resampled (again using weighted sampling).

In the case where there were more than 100 distinct label sets, some will be missing from the output. This may sound like a source of error, but an unbiased sampling algorithm is expected to introduce errors. Averaging the results across multiple windows will correct these errors. I like this sort of approach because it limits cardinality but conveys approximate full dimensionality.

I'm happy to leave the optional "sample_weight" out for now, because someone will have to demonstrate it in code. This will only start to matter to me when OTLP is more widely used, so it can wait.

Reply:
@MrAlias added a section talking about custom aggregations

@jmacd agreed with all of the points except for the last one, I think default behaviour might as well select exemplars statistically when possible (and use the parameter). I don't think we should go about defining statistical implementation details for the aggregations in this OTEP, but making sure we don't block it and aggregation descriptions encourage it is a reasonable goal, imo.

Reply:
Sounds good.